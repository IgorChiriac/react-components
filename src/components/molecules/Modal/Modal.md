### Summary

`<Modal />` is a flexible dialog prompt wrapping [`react-modal`](http://reactcommunity.org/react-modal/) with some basic styles.

Props are passed down to `react-modal` so if you want to customize it, please refer to its documentation.

#### Requisites ‚ö†Ô∏è

- For the accessibility, call `Modal.setAppElement('#rootElementId')` in the root component before any of the modals are open.
- Include `<Modal.Styles>` component (preferably in a top-most level component)

#### Tips üí¨

- Control the CSS `z-index` value applied to the modal through the `zIndex` prop on `<Modal.Styles />`
- Supply `onRequestClose` to close the modal when the user clicks on its overlay ( see [`react-modal` docs](http://reactcommunity.org/react-modal/examples/on_request_close.html) )

### Example

```jsx
import { Modal, Button, Heading } from '@zopauk/react-components';

// Call Modal.setAppElement ONCE in the top level level component and pass it a query selector indentifiying the root of your app.
Modal.setAppElement('#rsg-root');

function ModalDemo() {
  const [isOpen, update] = React.useState(false);
  const toggleModal = () => update(prevState => !prevState);

  return (
    <>
      <Button onClick={toggleModal} styling="link">
        Open Modal
      </Button>
      <Modal isOpen={isOpen} onRequestClose={toggleModal}>
        <Heading as="h2">Title of the modal</Heading>
        <p>text of the modal</p>
        <Button onClick={toggleModal} styling="link">
          Close Modal
        </Button>
      </Modal>
    </>
  );
}

<>
  <Modal.Styles zIndex={100} />
  <ModalDemo />
</>;
```
