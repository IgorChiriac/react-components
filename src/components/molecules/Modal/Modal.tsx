import React, { FC } from 'react';
import ReactModal from 'react-modal';
import ModalStyles from './ModalStyles/ModalStyles';
import Icon from '../../atoms/Icon/Icon';
import { faTimesCircle } from '@fortawesome/free-solid-svg-icons';
import { colors } from '../../../constants/colors';
import styled from 'styled-components';

export type ModalProps = ReactModal.Props & {
  showCloseButton?: boolean;
};

type ModalComponent = FC<ModalProps> & {
  Styles: typeof ModalStyles;
  setAppElement: typeof ReactModal.setAppElement;
};

const CrossIcon = styled(Icon).attrs({ color: colors.greyLight, variant: faTimesCircle })`
  position: absolute;
  cursor: pointer;
  right: 10px;
  top: 10px;
`;

const classNames = {
  afterOpen: 'zopa-modal--after-open',
  base: 'zopa-modal',
  beforeClose: 'zopa-modal--before-close',
};

const overlayClassNames = {
  afterOpen: 'zopa-modal-overlay--after-open',
  base: 'zopa-modal-overlay',
  beforeClose: 'zopa-modal-overlay--before-close',
};

const Modal: ModalComponent = ({ children, onRequestClose, showCloseButton = true, ...rest }) => (
  <ReactModal
    bodyOpenClassName="zopa-modal-body--open"
    portalClassName="zopa-modal-portal"
    className={classNames}
    overlayClassName={overlayClassNames}
    closeTimeoutMS={200}
    {...rest}
  >
    {showCloseButton && <CrossIcon onClick={onRequestClose} />}
    {children}
  </ReactModal>
);

Modal.Styles = ModalStyles;
Modal.setAppElement = ReactModal.setAppElement;

export default Modal;
