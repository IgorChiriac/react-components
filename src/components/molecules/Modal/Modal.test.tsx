import React from 'react';
import { render, waitForElement } from 'react-testing-library';
import Modal from './Modal';

describe('<Modal />', () => {
  it('render the opened Modal', async () => {
    Modal.setAppElement('body');
    const { getByText } = render(<Modal isOpen>text</Modal>);
    const modal = await waitForElement(() => getByText('text'));
    expect(modal).toMatchSnapshot();
  });
});
