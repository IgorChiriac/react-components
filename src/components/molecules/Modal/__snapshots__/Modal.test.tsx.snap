// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> allows to set global styles 1`] = `""`;

exports[`<Modal /> can render opened 1`] = `
<div
  class="zopa-modal zopa-modal--after-open"
  role="dialog"
  tabindex="-1"
>
  text
</div>
`;
